staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+)
+
+const (
+	// progressRequestPeriod determines period of requesting progress
+	// from etcd when there is a request waiting for watch cache to be fresh.
+	progressRequestPeriod = 100 * time.Millisecond
+)
+
+func newConditionalProgressRequester(requestProgress ProgressRequester, clock clock.WithTicker) *conditionalProgressRequester {
+	pr := &conditionalProgressRequester{
+		clock:           clock,
+		requestProgress: requestProgress,
+	}
+	mux := sync.Mutex{}
+	pr.cond = sync.NewCond(&mux)
+	mux.Lock()
+	return pr
+}
+
+// conditionalProgressRequester will request progress notification if there
+// is a request waiting for watch cache to be fresh.
+type conditionalProgressRequester struct {
+	clock           clock.WithTicker
+	requestProgress ProgressRequester
+
+	cond    *sync.Cond
+	waiting atomic.Int32
+}
+
+type ProgressRequester func(ctx context.Context) error
+
+func (pr *conditionalProgressRequester) Run(stopCh <-chan struct{}) {
+	ctx := wait.ContextForChannel(stopCh)
+	go func() {
+		<-stopCh
+		pr.cond.Signal()
+	}()
+	ticker := pr.clock.NewTicker(progressRequestPeriod)
+	defer ticker.Stop()
+	for {
+		for pr.waiting.Load() <= 0 {
+			select {
+			case <-stopCh:
+				return
+			default:
+			}
+			pr.cond.Wait()
```

Comment:
Is this racy? I didn't check how Signal/Wait works exactly.

Let's say that we:
- went through the select above (and stopCh was still opened)
- then stopCh was closed
- then gorotuine above sent Signal()
- then we come here and Wait()

Will it finish or not?

Reply:
It is.

Reply:
Ok, worked around it.

Reply:
[I can't remember the original version and if it worked, but it so might be worth getting back to one with int instead of atomic - but I'm not sure if it didn't have the exact same problem.]

Reply:
You are right and locks will solve the problem as they create a proper critical section. They guarantee that the worker loop is not between checking the condition and wait, but is always blocked on acquiring the lock or on wait.